drivers: gpio: gpio_pca953x: Adding input latch and interrupt mask

VuDangBP opened the pull request:
The gpio_pca953x gpio driver doesn't have
the input latch and interrupt mask
configuration which causes a lack of accessing
and using those features on an gpio expander
device. Fix it by adding input latch and
interrupt mask configurations in this driver.

nordicjm commented:
CC @rerickson1 long memory but this doesn't look valid for the BL5340 DVK port expander

rerickson1 commented on drivers/gpio/gpio_pca953x.c:
```diff
@@ -27,6 +27,8 @@ LOG_MODULE_REGISTER(pca953x, CONFIG_GPIO_LOG_LEVEL);
 #define PCA953X_INPUT_PORT		0x00
 #define PCA953X_OUTPUT_PORT		0x01
 #define PCA953X_CONFIGURATION		0x03
+#define REG_INPUT_LATCH_PORT0   0x42
+#define REG_INT_MASK_PORT0      0x45
```
spacing in consistency here

rerickson1 commented:
> CC @rerickson1 long memory but this doesn't look valid for the BL5340 DVK port expander

@nordicjm could you clarify your thinking here? Why wouldn't it work? I will start looking into this as well. Thanks!

nordicjm commented:
> > CC @rerickson1 long memory but this doesn't look valid for the BL5340 DVK port expander
> 
> @nordicjm could you clarify your thinking here? Why wouldn't it work? I will start looking into this as well. Thanks!

They don't exist in that part: https://www.ti.com/lit/ds/symlink/tca9538.pdf

VuDangBP commented on dts/bindings/gpio/ti,tca9538.yaml:
```diff
@@ -24,6 +24,19 @@ properties:
       Connection for the NINT signal. This signal is active-low when
       produced by tca9538 GPIO node.
 
+  input-latch:
+    type: int
+    default: 0
+    description: |
+      Setting property for Input Latch.
+
+  interrupt-mask:
+    type: int
+    default: 0
+    description: |
+      Setting property for Interrupt mask.
```
Hi @rerickson1, the `required` property is not set to true in both `input-latch` and `interrupt-mask` so I believe they are optional by default, please let me know if that isn't the correct way to do it.

VuDangBP commented on drivers/gpio/gpio_pca953x.c:
```diff
@@ -439,6 +443,12 @@ static int gpio_pca953x_init(const struct device *dev)
+		i2c_reg_write_byte_dt(&cfg->i2c, REG_INPUT_LATCH_PORT0, cfg->input_latch);
+
+		/* Interrupt mask config */
+		i2c_reg_write_byte_dt(&cfg->i2c, REG_INT_MASK_PORT0, cfg->interrupt_mask);
```
Thank you for pointing it out, has been updated.

VuDangBP commented:
Hi @rerickson1, as there is a new commit added to this PR, please review this PR. Thank you very much.

rerickson1 reviewed (changes requested):
Spaces missing here.

rerickson1 reviewed (changes requested):
Please squash all commits.

henrikbrixandersen commented on dts/bindings/gpio/ti,tca9538.yaml:
```diff
@@ -24,6 +24,17 @@ properties:
       Connection for the NINT signal. This signal is active-low when
       produced by tca9538 GPIO node.
 
+  input-latch:
+    type: int
+    description: |
+      Setting property for Input Latch.
+
+  interrupt-mask:
+    type: int
+    description: |
+      Setting property for Interrupt mask.
+
+
```
Please improve the descriptions here. It's difficult to know what to set these properties to.

VuDangBP replied:
Has been updated!

ycsin commented on drivers/gpio/gpio_pca953x.c:
```diff
@@ -467,7 +479,9 @@ static const struct gpio_driver_api api_table = {
 		},								\
 		.interrupt_enabled = DT_INST_NODE_HAS_PROP(n, nint_gpios),	\
-		.gpio_int = GPIO_DT_SPEC_INST_GET_OR(n, nint_gpios, {0}),	\
+		.gpio_int = GPIO_DT_SPEC_INST_GET_OR(n, nint_gpios, {0}),  \
```
alignment of `\` is off, resulting in unnecessary line change

VuDangBP replied:
Thank you for your comment, has been updated!

ycsin commented on drivers/gpio/gpio_pca953x.c:
```diff
@@ -467,7 +479,9 @@ static const struct gpio_driver_api api_table = {
-		.gpio_int = GPIO_DT_SPEC_INST_GET_OR(n, nint_gpios, {0}),	\
+		.gpio_int = GPIO_DT_SPEC_INST_GET_OR(n, nint_gpios, {0}),  \
+		.interrupt_mask = DT_INST_PROP_OR(n, interrupt_mask, -1),	\
+		.input_latch = DT_INST_PROP_OR(n, input_latch, -1)		\
```
clang-format will work properly for these if you add a comma at the end, i.e.

```c
.input_latch = DT_INST_PROP_OR(n, input_latch, -1),
```

VuDangBP replied:
Thank you for your comment, has been updated!

ycsin commented on dts/bindings/gpio/ti,tca9538.yaml:
```diff
@@ -24,6 +24,21 @@ properties:
+      enabling interrupts. Setting corresponding mask bits to logic
+      0 to enable the interrupts.
+
+
```
nits: unnecessary empty line

VuDangBP replied:
Thank you for the comment, will do it later if there's any comments requires code changes

fabiobaltieri replied:
still there, sorry don't mean to be harsh on nitpicks but if we start not caring the code base is going to become a giant mess, please fix

VuDangBP replied:
Yeah, has been updated!

ycsin reviewed (dismissed):
looks good, good to remove that empty line if there's another revision

VuDangBP committed:
drivers: gpio: gpio_pca953x: Adding input latch and interrupt mask

The gpio_pca953x gpio driver doesn't have
the input latch and interrupt mask
configuration which causes a lack of accessing
and using those features on an gpio expander
device. Fix it by adding input latch and
interrupt mask configurations in this driver.

Signed-off-by: Vudang Thaihai <vudang.thaihai@brillpower.com>

fabiobaltieri reviewed (approved):
Thanks for your patience :-)
